extensions/objects/templates/FilepathEditor.js

```diff
@@ -31,109 +31,118 @@ oppia.directive('filepathEditor', [
-        // Reset the component each time the value changes (e.g. if this is part
-        // of an editable list).
+        // Reset the component each time the value changes
+        // (e.g. if this is part of an editable list).
         $scope.$watch('$parent.value', function(newValue) {
-          $scope.localValue = {
-            label: newValue || ''
-          };
-          $scope.imageUploaderIsActive = false;
+          if (newValue) {
+            $scope.setSavedImageUrl(newValue, false);
+          }
         });
 
-        $scope.explorationId = explorationContextService.getExplorationId();
+        $scope.validate = function(state) {
+          return state.state === 'saved' &&
+                 state.savedUrl &&
+                 state.savedUrl.length > 0;
+        };
 
-        $scope.validate = function(localValue) {
-          return localValue.label && localValue.label.length > 0;
+        $scope.isNoImageUploaded = function() {
+          return !$scope.state || $scope.state.state === 'empty';
         };
 
-        $scope.$watch('localValue.label', function(newValue) {
-          if (newValue) {
-            alertsService.clearWarnings();
-            $scope.localValue = {
-              label: newValue
-            };
-            $scope.$parent.value = newValue;
-          }
-        });
+        $scope.isImageUploaded = function() {
+          return $scope.state && $scope.state.state === 'uploaded';
+        };
 
-        $scope.getPreviewUrl = function(filepath) {
-          var encodedFilepath = window.encodeURIComponent(filepath);
-          return $sce.trustAsResourceUrl(
-            '/imagehandler/' + $scope.explorationId + '/' + encodedFilepath);
+        $scope.isImageSaved = function() {
+          return $scope.state && $scope.state.state === 'saved';
         };
 
-        $scope.resetImageUploader = function() {
-          $scope.currentFile = null;
-          $scope.currentFilename = null;
-          $scope.imagePreview = null;
+        $scope.setEmptyState = function() {
+          $scope.state = {state: 'empty'};
         };
 
-        $scope.openImageUploader = function() {
-          $scope.resetImageUploader();
-          $scope.uploadWarning = null;
-          $scope.imageUploaderIsActive = true;
+        $scope.setUploadedFile = function(file) {
+          var reader = new FileReader();
+          reader.onload = function(e) {
+            var img = new Image();
+            img.onload = function() {
+              $scope.state = {
+                state: 'uploaded',
+                uploadedFile: file,
+                uploadedImageData: e.target.result,
+                originalWidth: this.naturalWidth,
+                originalHeight: this.naturalHeight
+              };
+              $scope.$apply();
+            };
+            img.src = e.target.result;
+          };
+          reader.readAsDataURL(file);
+        };
+
+        $scope.setSavedImageUrl = function(url, updateParent) {
```

Comment:
I think we should remove the second arg. In the 'false' case the caller can just set $scope.state (or $scope.data) directly.